From: Folia Compatibility Patch
Subject: [PATCH 0025] AlchemistInventoryClicks: Migrate to player entity scheduler and thread-safe collections

AE PATCH REASON: Uses legacy Bukkit scheduler inside Alchemist inventory close returning logic
AE PATCH FIX: Migrate to player entity scheduler and thread-safe collections
AE PATCH RISK: HIGH
AE PATCH EXCLUDES: net/advancedplugins/ae/features/alchemist/AlchemistInventoryClicks.class, net/advancedplugins/ae/features/alchemist/AlchemistInventoryClicks$*.class

---
diff --git a/net/advancedplugins/ae/features/alchemist/AlchemistInventoryClicks.java b/net/advancedplugins/ae/features/alchemist/AlchemistInventoryClicks.java
index 78d2c5e..997095a 100644
--- a/net/advancedplugins/ae/features/alchemist/AlchemistInventoryClicks.java
+++ b/net/advancedplugins/ae/features/alchemist/AlchemistInventoryClicks.java
@@ -32,232 +32,232 @@ import org.bukkit.inventory.meta.ItemMeta;
 import org.jetbrains.annotations.NotNull;
 
 public class AlchemistInventoryClicks implements Listener {
-   private static final List<UUID> pendingCloses = new ArrayList<>();
+   private static final List<UUID> pendingCloses = java.util.Collections.synchronizedList(new ArrayList<>());
    private final String _takeCurrency = "&c- %price%";
 
    @EventHandler
-   public void onClick(@NotNull InventoryClickEvent var1) {
-      if (AlchemistInventory.isAlchemistInventory(var1.getView())) {
-         if (AManager.isValid(var1.getCurrentItem())) {
-            var1.setCancelled(true);
-            int var2 = var1.getRawSlot();
-            Player var3 = (Player)var1.getWhoClicked();
-            ItemStack var4 = var1.getCurrentItem();
-            if (var4 == null) {
-               var4 = var1.getCursor();
+   public void onClick(@NotNull InventoryClickEvent event) {
+      if (AlchemistInventory.isAlchemistInventory(event.getView())) {
+         if (AManager.isValid(event.getCurrentItem())) {
+            event.setCancelled(true);
+            int rawSlot = event.getRawSlot();
+            Player player = (Player)event.getWhoClicked();
+            ItemStack clickedItem = event.getCurrentItem();
+            if (clickedItem == null) {
+               clickedItem = event.getCursor();
             }
 
-            if (var4 == null && var1.getClick().equals(ClickType.NUMBER_KEY)) {
-               var1.setCancelled(true);
-            } else if (var1.getClick() != ClickType.LEFT && var1.getClick() != ClickType.SHIFT_LEFT) {
-               var3.updateInventory();
+            if (clickedItem == null && event.getClick().equals(ClickType.NUMBER_KEY)) {
+               event.setCancelled(true);
+            } else if (event.getClick() != ClickType.LEFT && event.getClick() != ClickType.SHIFT_LEFT) {
+               player.updateInventory();
             } else {
-               int var5 = 0;
-               int var6 = var1.getView().getTopInventory().getSize() - 1;
-               if (var2 == AlchemistInventory.getConfirmItemSlot()) {
-                  if (var1.getInventory().getItem(AlchemistInventory.getPlayerFirstItemSlot()) != null
-                     || var1.getInventory().getItem(AlchemistInventory.getPlayerSecondItemSlot()) != null) {
-                     if (!NBTapi.contains("alchemist-worth", var4)) {
-                        var1.setCancelled(true);
+               int nextEmptySlot = 0;
+               int lastPlayerSlot = event.getView().getTopInventory().getSize() - 1;
+               if (rawSlot == AlchemistInventory.getConfirmItemSlot()) {
+                  if (event.getInventory().getItem(AlchemistInventory.getPlayerFirstItemSlot()) != null
+                     || event.getInventory().getItem(AlchemistInventory.getPlayerSecondItemSlot()) != null) {
+                     if (!NBTapi.contains("alchemist-worth", clickedItem)) {
+                        event.setCancelled(true);
                      } else {
-                        int var16 = Integer.parseInt(NBTapi.get("alchemist-worth", var4));
+                        int price = Integer.parseInt(NBTapi.get("alchemist-worth", clickedItem));
                         if (YamlFile.ALCHEMIST.contains("general.settings.price.economy")) {
-                           String var19 = YamlFile.ALCHEMIST.getString("general.settings.price.economy", "EXP");
-                           AdvancedEconomy var23 = EffectsHandler.getEconomyHandler().getEcon(var19);
-                           if (!var23.chargeUser(var3, var16)) {
-                              Lang.sendMessage(var3, "alchemist.cannot-afford", new String[0]);
+                           String currencyType = YamlFile.ALCHEMIST.getString("general.settings.price.economy", "EXP");
+                           AdvancedEconomy economy = EffectsHandler.getEconomyHandler().getEcon(currencyType);
+                           if (!economy.chargeUser(player, price)) {
+                              Lang.sendMessage(player, "alchemist.cannot-afford", new String[0]);
                               return;
                            }
 
-                           switch (var19) {
+                           switch (currencyType) {
                               case "EXP":
                                  Lang.sendMessageOrDefault(
-                                    var3, "alchemist.take-currency", "&c- %price%", new String[]{"%price%;" + AManager.formatNumber(var16) + " EXP"}
+                                    player, "alchemist.take-currency", "&c- %price%", new String[]{"%price%;" + AManager.formatNumber(price) + " EXP"}
                                  );
                                  break;
                               case "LEVEL":
                                  Lang.sendMessageOrDefault(
-                                    var3, "alchemist.take-currency", "&c- %price%", new String[]{"%price%;" + AManager.formatNumber(var16) + " LEVELS"}
+                                    player, "alchemist.take-currency", "&c- %price%", new String[]{"%price%;" + AManager.formatNumber(price) + " LEVELS"}
                                  );
                                  break;
                               case "MONEY":
                                  Lang.sendMessageOrDefault(
-                                    var3, "alchemist.take-currency", "&c- %price%", new String[]{"%price%;$" + AManager.formatNumber(var16)}
+                                    player, "alchemist.take-currency", "&c- %price%", new String[]{"%price%;$" + AManager.formatNumber(price)}
                                  );
                            }
                         } else if (YamlFile.ALCHEMIST.getBoolean("general.settings.price.experience", true)) {
-                           AdvancedEconomy var20 = EffectsHandler.getEconomyHandler().getEcon("EXP");
-                           if (!var20.chargeUser(var3, var16)) {
-                              Lang.sendMessage(var3, "alchemist.cannot-afford", new String[0]);
+                           AdvancedEconomy expEconomy = EffectsHandler.getEconomyHandler().getEcon("EXP");
+                           if (!expEconomy.chargeUser(player, price)) {
+                              Lang.sendMessage(player, "alchemist.cannot-afford", new String[0]);
                               return;
                            }
 
-                           var3.sendMessage("§c- " + AManager.formatNumber(var16) + " EXP");
+                           player.sendMessage("§c- " + AManager.formatNumber(price) + " EXP");
                         } else {
-                           if (Core.getEconomy().getBalance(var3) < var16) {
-                              Lang.sendMessage(var3, "alchemist.cannot-afford", new String[0]);
+                           if (Core.getEconomy().getBalance(player) < price) {
+                              Lang.sendMessage(player, "alchemist.cannot-afford", new String[0]);
                               return;
                            }
 
-                           Core.getEconomy().withdrawPlayer(var3, var16);
-                           var3.sendMessage("§c- $" + AManager.formatNumber(var16));
+                           Core.getEconomy().withdrawPlayer(player, price);
+                           player.sendMessage("§c- $" + AManager.formatNumber(price));
                         }
 
-                        pendingCloses.add(var3.getUniqueId());
-                        AlchemistTradeEvent var21 = new AlchemistTradeEvent(var3, var1.getInventory().getItem(AlchemistInventory.getPreviewItemSlot()));
-                        Bukkit.getPluginManager().callEvent(var21);
-                        AManager.giveItem(var3, new ItemStack[]{var1.getInventory().getItem(AlchemistInventory.getPreviewItemSlot())});
-                        AlchemistInventory.open(var3);
+                        pendingCloses.add(player.getUniqueId());
+                        AlchemistTradeEvent tradeEvent = new AlchemistTradeEvent(player, event.getInventory().getItem(AlchemistInventory.getPreviewItemSlot()));
+                        Bukkit.getPluginManager().callEvent(tradeEvent);
+                        AManager.giveItem(player, new ItemStack[]{event.getInventory().getItem(AlchemistInventory.getPreviewItemSlot())});
+                        AlchemistInventory.open(player);
                      }
                   }
                } else {
-                  if (var2 > var6 - 1) {
-                     if (!this.isAcceptable(var4)) {
-                        var1.setCancelled(true);
-                        if (var1.getCurrentItem() != null || var1.getCursor() != null) {
-                           Lang.sendMessage(var3, "alchemist.doesnt-accept", new String[0]);
+                  if (rawSlot > lastPlayerSlot - 1) {
+                     if (!this.isAcceptable(clickedItem)) {
+                        event.setCancelled(true);
+                        if (event.getCurrentItem() != null || event.getCursor() != null) {
+                           Lang.sendMessage(player, "alchemist.doesnt-accept", new String[0]);
                         }
 
                         return;
                      }
 
-                     if (var1.getInventory().getItem(AlchemistInventory.getPlayerFirstItemSlot()) != null
-                        && var1.getInventory().getItem(AlchemistInventory.getPlayerSecondItemSlot()) != null) {
-                        var1.setCancelled(true);
+                     if (event.getInventory().getItem(AlchemistInventory.getPlayerFirstItemSlot()) != null
+                        && event.getInventory().getItem(AlchemistInventory.getPlayerSecondItemSlot()) != null) {
+                        event.setCancelled(true);
                         return;
                      }
 
-                     int var7 = 0;
-                     if (var4.getAmount() > 1) {
-                        ItemStack var8 = var4.clone();
-                        var8.setAmount(var8.getAmount() - 1);
-                        var3.getInventory().setItem(var1.getSlot(), var8);
+                     int remainingAmount = 0;
+                     if (clickedItem.getAmount() > 1) {
+                        ItemStack clonedItem = clickedItem.clone();
+                        clonedItem.setAmount(clonedItem.getAmount() - 1);
+                        player.getInventory().setItem(event.getSlot(), clonedItem);
                      } else {
-                        var3.getInventory().setItem(var1.getSlot(), new ItemStack(Material.AIR));
+                        player.getInventory().setItem(event.getSlot(), new ItemStack(Material.AIR));
                      }
 
-                     if (var4.getAmount() > 1) {
-                        var7 = var4.getAmount() - 1;
-                        var4.setAmount(1);
+                     if (clickedItem.getAmount() > 1) {
+                        remainingAmount = clickedItem.getAmount() - 1;
+                        clickedItem.setAmount(1);
                      }
 
-                     if (var1.getInventory().getItem(AlchemistInventory.getPlayerFirstItemSlot()) == null) {
-                        var1.getInventory().setItem(AlchemistInventory.getPlayerFirstItemSlot(), var4);
-                        var5 = AlchemistInventory.getPlayerSecondItemSlot();
+                     if (event.getInventory().getItem(AlchemistInventory.getPlayerFirstItemSlot()) == null) {
+                        event.getInventory().setItem(AlchemistInventory.getPlayerFirstItemSlot(), clickedItem);
+                        nextEmptySlot = AlchemistInventory.getPlayerSecondItemSlot();
                      } else {
-                        if (var1.getInventory().getItem(AlchemistInventory.getPlayerSecondItemSlot()) != null) {
+                        if (event.getInventory().getItem(AlchemistInventory.getPlayerSecondItemSlot()) != null) {
                            return;
                         }
 
-                        var1.getInventory().setItem(AlchemistInventory.getPlayerSecondItemSlot(), var4);
-                        var5 = AlchemistInventory.getPlayerFirstItemSlot();
+                        event.getInventory().setItem(AlchemistInventory.getPlayerSecondItemSlot(), clickedItem);
+                        nextEmptySlot = AlchemistInventory.getPlayerFirstItemSlot();
                      }
 
-                     ItemStack var17 = var4.clone();
-                     var17.setAmount(var7);
-                  } else if (var2 == AlchemistInventory.getPlayerSecondItemSlot() || var2 == AlchemistInventory.getPlayerFirstItemSlot()) {
-                     var3.closeInventory();
-                     Lang.sendMessage(var3, "alchemist.trade-cancelled", new String[0]);
+                     ItemStack leftoverItem = clickedItem.clone();
+                     leftoverItem.setAmount(remainingAmount);
+                  } else if (rawSlot == AlchemistInventory.getPlayerSecondItemSlot() || rawSlot == AlchemistInventory.getPlayerFirstItemSlot()) {
+                     player.closeInventory();
+                     Lang.sendMessage(player, "alchemist.trade-cancelled", new String[0]);
                      return;
                   }
 
-                  if (var1.getInventory().getItem(var5) != null) {
-                     if (!this.isAcceptable(var4)) {
-                        var1.setCancelled(true);
-                        if (var1.getClickedInventory() != var1.getView().getTopInventory()) {
-                           Lang.sendMessage(var3, "alchemist.doesnt-accept", new String[0]);
+                  if (event.getInventory().getItem(nextEmptySlot) != null) {
+                     if (!this.isAcceptable(clickedItem)) {
+                        event.setCancelled(true);
+                        if (event.getClickedInventory() != event.getView().getTopInventory()) {
+                           Lang.sendMessage(player, "alchemist.doesnt-accept", new String[0]);
                         }
                      } else {
-                        ItemType var15 = this.getType(var1.getInventory().getItem(var5));
-                        ItemType var18 = this.getType(var4);
-                        if (!var15.equals(var18)) {
-                           if (var2 < var6) {
-                              if (var2 == AlchemistInventory.getPlayerFirstItemSlot()) {
-                                 AManager.giveItem(var3, new ItemStack[]{var4});
-                                 var1.getInventory().setItem(var2, new ItemStack(Material.AIR));
+                        ItemType placedType = this.getType(event.getInventory().getItem(nextEmptySlot));
+                        ItemType currentType = this.getType(clickedItem);
+                        if (!placedType.equals(currentType)) {
+                           if (rawSlot < lastPlayerSlot) {
+                              if (rawSlot == AlchemistInventory.getPlayerFirstItemSlot()) {
+                                 AManager.giveItem(player, new ItemStack[]{clickedItem});
+                                 event.getInventory().setItem(rawSlot, new ItemStack(Material.AIR));
                                  return;
                               }
 
-                              var1.setCancelled(true);
+                              event.setCancelled(true);
                            }
 
-                           var1.setCancelled(true);
-                        } else if (NBTapi.get("groupType", var4) != null
-                           && !NBTapi.get("groupType", var4).equalsIgnoreCase(NBTapi.get("groupType", var1.getInventory().getItem(var5)))) {
-                           var1.setCancelled(true);
-                           Lang.sendMessage(var3, "alchemist.doesnt-accept", new String[0]);
-                        } else if (AEAPI.isCustomEnchantBook(var4)
-                           && AEAPI.isCustomEnchantBook(var1.getInventory().getItem(var5))
-                           && !AEAPI.getBookEnchantment(var4).equalsIgnoreCase(AEAPI.getBookEnchantment(var1.getInventory().getItem(var5)))) {
-                           var1.setCancelled(true);
-                           Lang.sendMessage(var3, "alchemist.doesnt-accept", new String[0]);
+                           event.setCancelled(true);
+                        } else if (NBTapi.get("groupType", clickedItem) != null
+                           && !NBTapi.get("groupType", clickedItem).equalsIgnoreCase(NBTapi.get("groupType", event.getInventory().getItem(nextEmptySlot)))) {
+                           event.setCancelled(true);
+                           Lang.sendMessage(player, "alchemist.doesnt-accept", new String[0]);
+                        } else if (AEAPI.isCustomEnchantBook(clickedItem)
+                           && AEAPI.isCustomEnchantBook(event.getInventory().getItem(nextEmptySlot))
+                           && !AEAPI.getBookEnchantment(clickedItem).equalsIgnoreCase(AEAPI.getBookEnchantment(event.getInventory().getItem(nextEmptySlot)))) {
+                           event.setCancelled(true);
+                           Lang.sendMessage(player, "alchemist.doesnt-accept", new String[0]);
                         } else {
-                           int var9 = 0;
-                           switch (var15) {
+                           int tradePrice = 0;
+                           switch (placedType) {
                               case BOOK:
-                                 BookMaker var24 = new BookMaker(var4);
-                                 BookMaker var26 = new BookMaker(var1.getInventory().getItem(var5));
-                                 int var12 = AEnchants.matchEnchant(var24.getEnchant()).getHighestLevel();
-                                 if (var24.getLevel() + 1 > var12) {
-                                    var1.setCancelled(true);
-                                    Lang.sendMessage(var3, "alchemist.exceeds-max-level", new String[0]);
+                                 BookMaker placedBookMaker = new BookMaker(clickedItem);
+                                 BookMaker clickedBookMaker = new BookMaker(event.getInventory().getItem(nextEmptySlot));
+                                 int maxLevel = AEnchants.matchEnchant(placedBookMaker.getEnchant()).getHighestLevel();
+                                 if (placedBookMaker.getLevel() + 1 > maxLevel) {
+                                    event.setCancelled(true);
+                                    Lang.sendMessage(player, "alchemist.exceeds-max-level", new String[0]);
                                     return;
                                  }
 
-                                 if (var24.getLevel() != var26.getLevel()) {
-                                    var1.setCancelled(true);
-                                    Lang.sendMessage(var3, "alchemist.doesnt-accept", new String[0]);
+                                 if (placedBookMaker.getLevel() != clickedBookMaker.getLevel()) {
+                                    event.setCancelled(true);
+                                    Lang.sendMessage(player, "alchemist.doesnt-accept", new String[0]);
                                     return;
                                  }
 
-                                 var9 = 10 * (var24.getLevel() + var26.getLevel()) * ((var24.getSuccess() + var26.getSuccess()) / 2);
+                                 tradePrice = 10 * (placedBookMaker.getLevel() + clickedBookMaker.getLevel()) * ((placedBookMaker.getSuccess() + clickedBookMaker.getSuccess()) / 2);
                                  break;
                               case DUST:
-                                 int var10 = this.getSuccess(var4);
-                                 int var11 = this.getSuccess(var1.getInventory().getItem(var5));
-                                 var9 = 10 * (var10 + var11);
+                                 int placedDustSuccess = this.getSuccess(clickedItem);
+                                 int clickedDustSuccess = this.getSuccess(event.getInventory().getItem(nextEmptySlot));
+                                 tradePrice = 10 * (placedDustSuccess + clickedDustSuccess);
                                  break;
                               case OTHER:
-                                 var1.setCancelled(true);
+                                 event.setCancelled(true);
                                  return;
                            }
 
-                           var9 = YamlFile.ALCHEMIST.getBoolean("general.settings.price.let-plugin-count")
-                              ? var9
+                           tradePrice = YamlFile.ALCHEMIST.getBoolean("general.settings.price.let-plugin-count")
+                              ? tradePrice
                               : YamlFile.ALCHEMIST.getInt("general.settings.price.custom-price");
-                           var1.getInventory().setItem(AlchemistInventory.getConfirmItemSlot(), this.cloneConfirmItem(var3, var9));
-                           ItemStack var25 = null;
-                           switch (var15) {
+                           event.getInventory().setItem(AlchemistInventory.getConfirmItemSlot(), this.cloneConfirmItem(player, tradePrice));
+                           ItemStack resultItem = null;
+                           switch (placedType) {
                               case BOOK:
-                                 BookMaker var28 = new BookMaker(var4);
-                                 BookMaker var31 = new BookMaker(var1.getInventory().getItem(var5));
-                                 int var32 = var28.getSuccess() + var31.getSuccess();
-                                 int var33 = var28.getDestroy() + var31.getDestroy();
+                                 BookMaker placedBookMaker2 = new BookMaker(clickedItem);
+                                 BookMaker clickedBookMaker2 = new BookMaker(event.getInventory().getItem(nextEmptySlot));
+                                 int combinedSuccess = placedBookMaker2.getSuccess() + clickedBookMaker2.getSuccess();
+                                 int combinedDestroy = placedBookMaker2.getDestroy() + clickedBookMaker2.getDestroy();
                                  if (YamlFile.ALCHEMIST.getBoolean("general.settings.books.add-up-chances")) {
-                                    var32 /= 2;
-                                    var33 /= 2;
+                                    combinedSuccess /= 2;
+                                    combinedDestroy /= 2;
                                  } else if (YamlFile.ALCHEMIST.getBoolean("general.settings.books.pre-set-chances.enabled")) {
-                                    var32 = YamlFile.ALCHEMIST.getInt("general.settings.books.pre-set-chances.success");
-                                    var33 = YamlFile.ALCHEMIST.getInt("general.settings.books.pre-set-chances.destroy");
+                                    combinedSuccess = YamlFile.ALCHEMIST.getInt("general.settings.books.pre-set-chances.success");
+                                    combinedDestroy = YamlFile.ALCHEMIST.getInt("general.settings.books.pre-set-chances.destroy");
                                  }
 
-                                 var25 = BookMaker.createABook(
-                                    new APloppableEnchantment(var32, var33, AEnchants.matchEnchant(var28.getEnchant()), var28.getLevel() + 1, var3)
+                                 resultItem = BookMaker.createABook(
+                                    new APloppableEnchantment(combinedSuccess, combinedDestroy, AEnchants.matchEnchant(placedBookMaker2.getEnchant()), placedBookMaker2.getLevel() + 1, player)
                                  );
                                  break;
                               case DUST:
-                                 int var27 = this.getSuccess(var4);
-                                 int var30 = this.getSuccess(var1.getInventory().getItem(var5));
-                                 int var13 = var30 + var27;
-                                 if (var13 > 100) {
-                                    var13 = 100;
+                                 int placedDustSuccess2 = this.getSuccess(clickedItem);
+                                 int clickedDustSuccess2 = this.getSuccess(event.getInventory().getItem(nextEmptySlot));
+                                 int combinedSuccessRate = placedDustSuccess2 + clickedDustSuccess2;
+                                 if (combinedSuccessRate > 100) {
+                                    combinedSuccessRate = 100;
                                  }
 
-                                 AdvancedGroup var14 = AdvancedGroup.matchGroup(NBTapi.get("groupType", var4));
-                                 var25 = TinkererItems.magicDustNonRandom(var14, var13);
+                                 AdvancedGroup enchantGroup = AdvancedGroup.matchGroup(NBTapi.get("groupType", clickedItem));
+                                 resultItem = TinkererItems.magicDustNonRandom(enchantGroup, combinedSuccessRate);
                            }
 
-                           var1.getInventory().setItem(AlchemistInventory.getPreviewItemSlot(), var25);
+                           event.getInventory().setItem(AlchemistInventory.getPreviewItemSlot(), resultItem);
                         }
                      }
                   }
@@ -267,118 +267,118 @@ public class AlchemistInventoryClicks implements Listener {
       }
    }
 
-   private boolean isAcceptable(ItemStack var1) {
-      if (AEAPI.isCustomEnchantBook(var1)) {
+   private boolean isAcceptable(ItemStack item) {
+      if (AEAPI.isCustomEnchantBook(item)) {
          return YamlFile.ALCHEMIST.getBoolean("general.settings.combine-books");
       } else {
-         return NBTapi.contains("magic", var1) ? YamlFile.ALCHEMIST.getBoolean("general.settings.combine-dust") : false;
+         return NBTapi.contains("magic", item) ? YamlFile.ALCHEMIST.getBoolean("general.settings.combine-dust") : false;
       }
    }
 
-   private ItemType getType(ItemStack var1) {
-      if (AEAPI.isCustomEnchantBook(var1)) {
+   private ItemType getType(ItemStack item) {
+      if (AEAPI.isCustomEnchantBook(item)) {
          return ItemType.BOOK;
       } else {
-         return NBTapi.contains("magic", var1) ? ItemType.DUST : ItemType.OTHER;
+         return NBTapi.contains("magic", item) ? ItemType.DUST : ItemType.OTHER;
       }
    }
 
-   private int getSuccess(ItemStack var1) {
-      return Integer.parseInt(NBTapi.get("magic", var1));
+   private int getSuccess(ItemStack item) {
+      return Integer.parseInt(NBTapi.get("magic", item));
    }
 
-   private void returnItems(Player var1, Inventory var2, boolean var3) {
-      Bukkit.getScheduler().runTaskLater(Core.getInstance(), () -> {
-         if (var1.isOnline()) {
-            if (AlchemistInventory.isAlchemistInventory(var1.getOpenInventory())) {
-               ItemStack var3x = var2.getItem(AlchemistInventory.getPlayerFirstItemSlot());
-               ItemStack var4 = var2.getItem(AlchemistInventory.getPlayerSecondItemSlot());
-               if (!(var1.getHealth() <= 0.0) && var1.getType().isAlive() && var1.isValid()) {
-                  if (var3x != null) {
-                     AManager.giveItem(var1, new ItemStack[]{var3x});
+   private void returnItems(Player player, Inventory inventory, boolean reopenInventory) {
+      player.getScheduler().runDelayed(Core.getInstance(), task -> {
+         if (player.isOnline()) {
+            if (AlchemistInventory.isAlchemistInventory(player.getOpenInventory())) {
+               ItemStack firstItem = inventory.getItem(AlchemistInventory.getPlayerFirstItemSlot());
+               ItemStack secondItem = inventory.getItem(AlchemistInventory.getPlayerSecondItemSlot());
+               if (!(player.getHealth() <= 0.0) && player.getType().isAlive() && player.isValid()) {
+                  if (firstItem != null) {
+                     AManager.giveItem(player, new ItemStack[]{firstItem});
                   }
 
-                  if (var4 != null) {
-                     AManager.giveItem(var1, new ItemStack[]{var4});
+                  if (secondItem != null) {
+                     AManager.giveItem(player, new ItemStack[]{secondItem});
                   }
 
-                  if (var3) {
-                     pendingCloses.add(var1.getUniqueId());
-                     AlchemistInventory.open(var1);
+                  if (reopenInventory) {
+                     pendingCloses.add(player.getUniqueId());
+                     AlchemistInventory.open(player);
                   }
                } else {
-                  if (var3x != null) {
-                     var1.getWorld().dropItemNaturally(var1.getLocation(), var3x);
+                  if (firstItem != null) {
+                     player.getWorld().dropItemNaturally(player.getLocation(), firstItem);
                   }
 
-                  if (var4 != null) {
-                     var1.getWorld().dropItemNaturally(var1.getLocation(), var4);
+                  if (secondItem != null) {
+                     player.getWorld().dropItemNaturally(player.getLocation(), secondItem);
                   }
                }
             }
          }
-      }, 1L);
+      }, null, 1L);
    }
 
    @EventHandler
-   public void onClose(InventoryCloseEvent var1) {
+   public void onClose(InventoryCloseEvent event) {
       try {
-         if (!AlchemistInventory.isAlchemistInventory(var1.getView())) {
+         if (!AlchemistInventory.isAlchemistInventory(event.getView())) {
             return;
          }
 
-         if (pendingCloses.contains(var1.getPlayer().getUniqueId())) {
-            pendingCloses.remove(var1.getPlayer().getUniqueId());
+         if (pendingCloses.contains(event.getPlayer().getUniqueId())) {
+            pendingCloses.remove(event.getPlayer().getUniqueId());
             return;
          }
 
-         Player var2 = (Player)var1.getPlayer();
-         if (!var2.isOnline()) {
+         Player player = (Player)event.getPlayer();
+         if (!player.isOnline()) {
             return;
          }
 
-         ItemStack var3 = var1.getInventory().getItem(AlchemistInventory.getPlayerFirstItemSlot());
-         ItemStack var4 = var1.getInventory().getItem(AlchemistInventory.getPlayerSecondItemSlot());
-         if (var2.getHealth() <= 0.0 || !var2.getType().isAlive()) {
-            if (var3 != null) {
-               var2.getWorld().dropItemNaturally(var2.getLocation(), var3);
+         ItemStack firstItem = event.getInventory().getItem(AlchemistInventory.getPlayerFirstItemSlot());
+         ItemStack secondItem = event.getInventory().getItem(AlchemistInventory.getPlayerSecondItemSlot());
+         if (player.getHealth() <= 0.0 || !player.getType().isAlive()) {
+            if (firstItem != null) {
+               player.getWorld().dropItemNaturally(player.getLocation(), firstItem);
             }
 
-            if (var4 != null) {
-               var2.getWorld().dropItemNaturally(var2.getLocation(), var4);
+            if (secondItem != null) {
+               player.getWorld().dropItemNaturally(player.getLocation(), secondItem);
             }
 
             return;
          }
 
-         if (var3 != null) {
-            AManager.giveItem(var2, new ItemStack[]{var3});
+         if (firstItem != null) {
+            AManager.giveItem(player, new ItemStack[]{firstItem});
          }
 
-         if (var4 != null) {
-            AManager.giveItem(var2, new ItemStack[]{var4});
+         if (secondItem != null) {
+            AManager.giveItem(player, new ItemStack[]{secondItem});
          }
-      } catch (Exception var5) {
+      } catch (Exception exception) {
       }
    }
 
-   private ItemStack cloneConfirmItem(Player var1, int var2) {
-      ItemStack var3 = AlchemistInventory.getConfirmItem().clone();
-      ItemMeta var4 = var3.getItemMeta();
-      ArrayList var5 = new ArrayList();
-      boolean var6 = Bukkit.getPluginManager().isPluginEnabled("PlaceholderAPI");
+   private ItemStack cloneConfirmItem(Player player, int price) {
+      ItemStack confirmItem = AlchemistInventory.getConfirmItem().clone();
+      ItemMeta itemMeta = confirmItem.getItemMeta();
+      ArrayList loreLines = new ArrayList();
+      boolean placeholderApiEnabled = Bukkit.getPluginManager().isPluginEnabled("PlaceholderAPI");
 
-      for (int var7 = 0; var7 < var4.getLore().size(); var7++) {
-         String var8 = ((String)var4.getLore().get(var7)).replaceAll("%price%", Integer.toString(var2));
-         if (var6) {
-            var8 = PlaceholderAPI.setPlaceholders(var1, var8);
+      for (int index = 0; index < itemMeta.getLore().size(); index++) {
+         String line = ((String)itemMeta.getLore().get(index)).replaceAll("%price%", Integer.toString(price));
+         if (placeholderApiEnabled) {
+            line = PlaceholderAPI.setPlaceholders(player, line);
          }
 
-         var5.add(var7, var8);
+         loreLines.add(index, line);
       }
 
-      var4.setLore(var5);
-      var3.setItemMeta(var4);
-      return NBTapi.addNBTTag("alchemist-worth", Integer.toString(var2), var3);
+      itemMeta.setLore(loreLines);
+      confirmItem.setItemMeta(itemMeta);
+      return NBTapi.addNBTTag("alchemist-worth", Integer.toString(price), confirmItem);
    }
 }

